FK_solver/src/FK_solver.cpp

```diff
@@ -0,0 +1,341 @@
+    dof_handler.distribute_dofs(*fe);
+
+    locally_owned_dofs = dof_handler.locally_owned_dofs();
+    DoFTools::extract_locally_relevant_dofs(dof_handler, locally_relevant_dofs);
+
+    pcout << "  Number of DoFs = " << dof_handler.n_dofs() << std::endl;
+  }
+
+  pcout << "-----------------------------------------------" << std::endl;
+
+  // Initialize the linear system.
+  {
+    pcout << "Initializing the linear system" << std::endl;
+
+    pcout << "  Initializing the sparsity pattern" << std::endl;
+
+    TrilinosWrappers::SparsityPattern sparsity(locally_owned_dofs,
+                                               MPI_COMM_WORLD);
+    DoFTools::make_sparsity_pattern(dof_handler, sparsity);
+    sparsity.compress();
+
+    pcout << "  Initializing the matrices" << std::endl;
+    jacobian_matrix.reinit(sparsity);
+
+    pcout << "  Initializing the system right-hand side" << std::endl;
+    residual_vector.reinit(locally_owned_dofs, MPI_COMM_WORLD);
+    pcout << "  Initializing the solution vector" << std::endl;
+    solution_owned.reinit(locally_owned_dofs, MPI_COMM_WORLD);
+    delta_owned.reinit(locally_owned_dofs, MPI_COMM_WORLD);
+
+    solution.reinit(locally_owned_dofs, locally_relevant_dofs, MPI_COMM_WORLD);
+    solution_old = solution;
+  }
+}
+
+void
+FK_solver::assemble_system()
+{
+  const unsigned int dofs_per_cell = fe->dofs_per_cell;
+  const unsigned int n_q           = quadrature->size();
+
+  FEValues<dim> fe_values(*fe,
+                          *quadrature,
+                          update_values | update_gradients |
+                            update_quadrature_points | update_JxW_values);
+
+  FullMatrix<double> cell_matrix(dofs_per_cell, dofs_per_cell);
+  Vector<double>     cell_residual(dofs_per_cell);
+
+  std::vector<types::global_dof_index> dof_indices(dofs_per_cell);
+
+  jacobian_matrix = 0.0;
+  residual_vector = 0.0;
+
+  // Value and gradient of the solution on current cell.
+  std::vector<double>         solution_loc(n_q);
+  std::vector<Tensor<1, dim>> solution_gradient_loc(n_q);
+
+  // Value of the solution at previous timestep (un) on current cell.
+  std::vector<double> solution_old_loc(n_q);
+
+  forcing_term.set_time(time);
+
+  for (const auto &cell : dof_handler.active_cell_iterators())
+    {
+      if (!cell->is_locally_owned())
+        continue;
+
+      fe_values.reinit(cell);
+
+      cell_matrix   = 0.0;
+      cell_residual = 0.0;
+
+      fe_values.get_function_values(solution, solution_loc);
+      fe_values.get_function_gradients(solution, solution_gradient_loc);
+      fe_values.get_function_values(solution_old, solution_old_loc);
+
+      for (unsigned int q = 0; q < n_q; ++q) {
+    // Evaluate coefficients on this quadrature node.
+    const double alpha_loc = alpha.value(fe_values.quadrature_point(q));
+
+    // Define the matrix to store D values and populate it.
+    FullMatrix<double> D_matrix(dim, dim);
+    D.matrix_value(fe_values.quadrature_point(q), D_matrix);
```

Comment:
Since you are already not using the `Function` interface, you can change this to an interface that uses the return value rather than input parameters:
```cpp
const FullMatrix<double> D_matrix = D.matrix_value(fe_values.quadrature_point(q));
```
This is generally more modern and readable, and thanks to return value optimization it should have zero overhead.

If instead you want to be sure not to allocate (and then de-allocate) a new `FullMatrix` for every quadrature node, you should keep the interface as it is, but instead define `D_matrix` outside the loop on elements, and reuse it for every iteration.